src/renderer/hooks/usePaneNodeSearchMenu.tsx

```diff
@@ -0,0 +1,299 @@
+import { Node, OnConnectStartParams, useReactFlow } from 'react-flow-renderer';
+import { useContext } from 'use-context-selector';
+import { NodeData, NodeSchema } from '../../common/common-types';
+import { createUniqueId, parseHandle } from '../../common/util';
+import { IconFactory } from '../components/CustomIcons';
+import { ContextMenuContext } from '../contexts/ContextMenuContext';
+import { GlobalContext, GlobalVolatileContext, NodeProto } from '../contexts/GlobalNodeState';
+import getNodeAccentColors from '../helpers/getNodeAccentColors';
+import { getMatchingNodes, getNodesByCategory } from '../helpers/nodeSearchFuncs';
+import { useContextMenu } from './useContextMenu';
+
+interface UsePaneNodeSearchMenuValue {
+    readonly onConnectStart: (event: React.MouseEvent, handle: OnConnectStartParams) => void;
+    readonly onConnectStop: (event: MouseEvent) => void;
+    readonly onPaneContextMenu: (event: React.MouseEvent) => void;
+}
+
+export const usePaneNodeSearchMenu = (
+    wrapperRef: React.RefObject<HTMLDivElement>
+): UsePaneNodeSearchMenuValue => {
+    const { createNode, createConnection } = useContext(GlobalVolatileContext);
+    const { closeContextMenu } = useContext(ContextMenuContext);
+    const { schemata } = useContext(GlobalContext);
+
+    const [connectingFrom, setConnectingFrom] = useState<OnConnectStartParams | null>();
+    const [connectingFromType, setConnectingFromType] = useState<string | null>();
+    const [isStoppedOnPane, setIsStoppedOnPane] = useState<boolean>(false);
+    const { getNode, project } = useReactFlow();
+
+    const [searchQuery, setSearchQuery] = useState<string>('');
+    const matchingNodes = useMemo(
+        () =>
+            getMatchingNodes(searchQuery, schemata.schemata).filter((node) => {
+                if (!connectingFrom || !connectingFromType) {
+                    return true;
+                }
+                if (connectingFrom.handleType === 'source') {
+                    return node.inputs.some((input) => {
+                        return connectingFromType === input.type && input.hasHandle;
+                    });
+                }
+                if (connectingFrom.handleType === 'target') {
+                    return node.outputs.some((output) => {
+                        return connectingFromType === output.type;
+                    });
+                }
+                log.error(`Unknown handle type: ${connectingFrom.handleType!}`);
+                return true;
+            }),
+        [connectingFrom, connectingFromType, searchQuery, schemata.schemata]
+    );
+    const byCategories = useMemo(() => getNodesByCategory(matchingNodes), [matchingNodes]);
+    const menuRef = useRef<HTMLDivElement>(null);
+
+    useEffect(() => {
+        setSearchQuery('');
+    }, [connectingFrom]);
+
+    const onPaneContextMenuNodeClick = useCallback(
+        (node: NodeSchema, position: { x: number; y: number }) => {
```

Comment:
You are using this type below as well. Could you please give it a name?

Also, since it's used as React state, could you please make the fields readonly?

Reply:
What do you mean? 

```ts
interface Position {
    readonly x: number;
    readonly y: number;
}
``` 
works fine.

Reply:
Ah, I was trying to do `x: readonly number;`